modules/apps/portal-store/portal-store-s3/src/test/java/com/liferay/portal/store/s3/S3StoreUnitTest.java

```diff
@@ -0,0 +1,262 @@
+		_mockProxy(null, null);
+
+		Assert.assertTrue(_testProxy(null, null));
+	}
+
+	@Test
+	public void testProxyAuthentication() throws Exception {
+		String proxyUserName = RandomTestUtil.randomString();
+		String proxyPassword = RandomTestUtil.randomString();
+
+		_mockProxy(proxyUserName, proxyPassword);
+
+		Assert.assertTrue(_testProxy(proxyUserName, proxyPassword));
+	}
+
+	@Test
+	public void testProxyAuthenticationFailed() throws Exception {
+		String proxyUserName = RandomTestUtil.randomString();
+		String proxyPassword = RandomTestUtil.randomString();
+
+		_mockProxy(proxyUserName, proxyPassword);
+
+		Assert.assertFalse(_testProxy(proxyUserName, proxyPassword + "1"));
+	}
+
+	private void _mockProxy(String proxyUserName, String proxyPassword) {
+		Mockito.when(
+			_s3StoreConfiguration.proxyHost()
+		).thenReturn(
+			_INET_SOCKET_ADDRESS.getHostString()
+		);
+
+		Mockito.when(
+			_s3StoreConfiguration.proxyPort()
+		).thenReturn(
+			_INET_SOCKET_ADDRESS.getPort()
+		);
+
+		if (Validator.isNotNull(proxyUserName)) {
+			Mockito.when(
+				_s3StoreConfiguration.proxyAuthType()
+			).thenReturn(
+				"username-password"
+			);
+
+			Mockito.when(
+				_s3StoreConfiguration.proxyPassword()
+			).thenReturn(
+				proxyPassword
+			);
+
+			Mockito.when(
+				_s3StoreConfiguration.proxyUsername()
+			).thenReturn(
+				proxyUserName
+			);
+		}
+		else {
+			Mockito.when(
+				_s3StoreConfiguration.proxyAuthType()
+			).thenReturn(
+				"none"
+			);
+		}
+	}
+
+	private boolean _testProxy(String proxyUserName, String proxyPassword)
+		throws Exception {
+
+		AtomicBoolean proxyHit = new AtomicBoolean(false);
+
+		HttpProxyServerBootstrap httpProxyServerBootstrap =
+			DefaultHttpProxyServer.bootstrap();
+
+		httpProxyServerBootstrap.withAddress(_INET_SOCKET_ADDRESS);
+		httpProxyServerBootstrap.withFiltersSource(
+			new HttpFiltersSourceAdapter() {
+
+				@Override
+				public HttpFilters filterRequest(HttpRequest httpRequest) {
+					proxyHit.set(true);
+
+					return super.filterRequest(httpRequest);
+				}
+
+			});
+
+		if (Validator.isNotNull(proxyUserName)) {
+			httpProxyServerBootstrap.withProxyAuthenticator(
+				new ProxyAuthenticator() {
+
+					@Override
+					public boolean authenticate(
+						String userName, String password) {
+
+						if (Objects.equals(userName, proxyUserName) &&
+							Objects.equals(password, proxyPassword)) {
+
+							return true;
+						}
+
+						return false;
+					}
+
+					@Override
+					public String getRealm() {
+						return null;
+					}
+
+				});
+		}
+
+		HttpProxyServer httpProxyServer = httpProxyServerBootstrap.start();
+
+		try {
+			S3Store s3Store = new S3Store();
+
+			s3Store.activate(Collections.emptyMap());
+
+			try {
+				s3Store.getFileSize(
+					RandomTestUtil.randomLong(), RandomTestUtil.randomLong(),
+					RandomTestUtil.randomString(), Store.VERSION_DEFAULT);
+
+				return proxyHit.get();
+			}
+			catch (Exception exception) {
+				return proxyHit.get();
```

Comment:
@kevhlee So, `s3Store.getFileSize` will hit the proxy. For `testProxy` and `testProxyAuthentication` the logic should return at line 241 in the try-block. But for `testProxyAuthenticationFailed` it just returns at this line.

The problem here, which also caused some confusion before, is that we don't assert the exception. So we don't know whether the exception is an expected one (auth failure of the proxy), or an unexpected one (the credential thing).

Please add assertion for the exception case.

Reply:
@dantewang Actually, line 241 should never be hit in all the proxy test cases (I've done this intentionally). Technically, an exception should always occur either because the S3 client failed to authenticate with the proxy or the client does not properly connect to a S3 bucket (the configuration is invalid intentionally so we don't have to deal with an actual S3 bucket).

I guess to make the logic clearer, it may be better to replace line 241 with `Assert.fail()` instead. Additionally, I can also add assertions for the exception to make sure it is expected. Anything else I should do to make the logic clearer?